sql/moz-fx-data-shared-prod/braze_external/changed_users_v1/query.sql

```diff
@@ -0,0 +1,52 @@
+SELECT
+  COALESCE(current_users.external_id, previous_users.external_id) AS external_id,
+  CASE
+    WHEN current_users.external_id IS NULL
+      THEN 'Deleted'
+    WHEN NOT (
+        current_users.email = previous_users.email
+        AND current_users.mailing_country = previous_users.mailing_country
+        AND current_users.email_subscribe = previous_users.email_subscribe
+        AND current_users.basket_token = previous_users.basket_token
+        AND current_users.email_lang = previous_users.email_lang
+        AND current_users.create_timestamp = previous_users.create_timestamp
+        AND current_users.update_timestamp = previous_users.update_timestamp
+        AND current_users.fxa_id_sha256 = previous_users.fxa_id_sha256
+        AND current_users.has_fxa = previous_users.has_fxa
+        AND current_users.fxa_primary_email = previous_users.fxa_primary_email
+        AND current_users.fxa_lang = previous_users.fxa_lang
+        AND current_users.first_service = previous_users.first_service
+      )
+      THEN 'Changed'
+  END AS status,
+  COALESCE(current_users.email, previous_users.email) AS email,
+  COALESCE(current_users.mailing_country, previous_users.mailing_country) AS mailing_country,
+  COALESCE(current_users.email_subscribe, previous_users.email_subscribe) AS email_subscribe,
+  COALESCE(current_users.basket_token, previous_users.basket_token) AS basket_token,
+  COALESCE(current_users.email_lang, previous_users.email_lang) AS email_lang,
+  COALESCE(current_users.create_timestamp, previous_users.create_timestamp) AS create_timestamp,
+  COALESCE(current_users.update_timestamp, previous_users.update_timestamp) AS update_timestamp,
+  COALESCE(current_users.has_fxa, previous_users.has_fxa) AS has_fxa,
+  COALESCE(current_users.fxa_primary_email, previous_users.fxa_primary_email) AS fxa_primary_email,
+  COALESCE(current_users.fxa_lang, previous_users.fxa_lang) AS fxa_lang,
+  COALESCE(current_users.first_service, previous_users.first_service) AS first_service
+FROM
+  `moz-fx-data-shared-prod.braze_derived.users_v1` current_users
+FULL OUTER JOIN
+  `moz-fx-data-shared-prod.braze_external.users_previous_day_snapshot_v1` previous_users
+  ON current_users.external_id = previous_users.external_id
+WHERE
+  current_users.external_id IS NULL  -- deleted rows
```

Comment:
Does this make it equivalent to a right join?  It's fine either way if it is.  I just want to make sure I understand

Reply:
The specific part of the query you're asking about, where it filters with current_users.external_id IS NULL, is focusing on identifying rows that appear in the previous_users (right table) but not in the current_users (left table). This behaves similar to what we would expect from a right join—specifically, identifying "deleted" rows that are present in the right table and absent from the left table

However, it's important to note that this is just a part of the larger query that uses a FULL OUTER JOIN, returning all rows from both current_users and previous_users, where:

- rows match across both tables on external_id
- unmatched rows from both tables are included with NULLs for the missing side

The condition current_users.external_id IS NULL does make this segment behave similarly to a right join, but only in the context of filtering for deleted users. It doesn't affect the overall join type of the query, which also considers rows that might be new (present in current_users but absent in previous_users) and other discrepancies beyond just the presence of an external_id